src/toxcore/dht/server/mod.rs

```diff
@@ -184,25 +173,50 @@ impl Server {
 
         Box::new(pings_stream.for_each(|()| Ok(())))
     }
-    /// send NodesRequest to random peer every 20 seconds
-    pub fn send_nodes_req(&self, friend_pk: PublicKey) -> IoFuture<()> {
-        let mut state = self.state.write();
-        if let Some(peer) = state.kbucket.get_random_node() {
-            let client = state.peers_cache.entry(peer.pk).or_insert_with(ClientData::new);
 
-            let payload = NodesRequestPayload {
-                pk: friend_pk,
-                id: client.new_ping_id(),
-            };
-            let nodes_req = DhtPacket::NodesRequest(NodesRequest::new(
-                &precompute(&peer.pk, &self.sk),
-                &self.pk,
-                payload
-            ));
-            self.send_to(peer.saddr, nodes_req)
-        } else {
-            Box::new(future::ok(()))
+    /// Send NodesRequest to random peer every 20 seconds
+    pub fn periodical_nodes_req(&self) -> IoFuture<()> {
+        let peer = match self.get_random_node() {
+            None => {
+                return Box::new( future::ok(()))
+            },
+            Some(peer) => peer,
+        };
+
+        self.send_nodes_req(peer)
+    }
+    
+    // Get random node from kbucket
+    fn get_random_node(&self) -> Option<PackedNode> {
+        let state = self.state.read();
+        state.kbucket.get_random_node()
+    }
+
+    // Send NodesRequest to peer
+    fn send_nodes_req(&self, target_peer: PackedNode) -> IoFuture<()> {
+        // Check if packet is going to be sent to ourself.
+        if self.pk == target_peer.pk {
+            return Box::new(
+                future::err(
+                    Error::new(ErrorKind::Other, "friend's pk is mine")
+                )
+            )
         }
+
+        let mut state = self.state.write();
```

Comment:
Maybe get random node here from kbucket? In this case you won't acquire lock twice. Anyway this fuction is used for sending nodes request to random peers only.

Reply:
Sending nodes request to random peers is periodical_nodes_req().
send_nodes_req(target_peer) gets target_peer as a argument.
In c-toxcore, NodesRequest is used for ping peers.
send_nodes_req() will be used to ping peers.

Reply:
ok then.